src/mlpack/core/data/load_impl.hpp

```diff
@@ -20,376 +21,300 @@
+bool LoadDense(const std::string& filename,
+               MatType& matrix,
+               TextOptions& opts,
+               std::fstream& stream)
 {
-  Timer::Start("loading_data");
-
-  // Get the extension.
-  std::string extension = Extension(filename);
+  bool success;
+  if (opts.Format() != FileType::RawBinary)
+    Log::Info << "Loading '" << filename << "' as " 
+        << opts.FileTypeToString() << ".  " << std::flush;
 
-  // Catch nonexistent files by opening the stream ourselves.
-  std::fstream stream;
-#ifdef  _WIN32 // Always open in binary mode on Windows.
-  stream.open(filename.c_str(), std::fstream::in | std::fstream::binary);
-#else
-  stream.open(filename.c_str(), std::fstream::in);
-#endif
-  if (!stream.is_open())
+  // We can't use the stream if the type is HDF5.
+  if (opts.Format() == FileType::HDF5Binary)
   {
-    Timer::Stop("loading_data");
-    if (fatal)
-      Log::Fatal << "Cannot open file '" << filename << "'. " << std::endl;
-    else
-      Log::Warn << "Cannot open file '" << filename << "'; load failed."
-          << std::endl;
-
-    return false;
+    success = LoadHDF5(filename, matrix, opts);
   }
+  else if (opts.Format() == FileType::CSVASCII)
+  {
+    success = LoadCSVASCII(filename, matrix, opts);
 
-  FileType loadType = inputLoadType;
-  std::string stringType;
-  if (inputLoadType == FileType::AutoDetect)
+    if (matrix.col(0).is_zero())
+      Log::Warn << "First column is zeros, if dataset contains headers, "
+                   "please use HasHeaders() option." << std::endl;
+  }
+  else
   {
-    // Attempt to auto-detect the type from the given file.
-    loadType = AutoDetect(stream, filename);
-    // Provide an error if we don't know the type, or if the type can't be used
-    // to load sparse matrices.
-    if (loadType == FileType::FileTypeUnknown)
-    {
-      if (fatal)
-        Log::Fatal << "Unable to detect type of '" << filename << "'; "
-            << "incorrect extension?" << std::endl;
-      else
-        Log::Warn << "Unable to detect type of '" << filename << "'; load "
-            << "failed. Incorrect extension?" << std::endl;
+    if (opts.Format() == FileType::RawBinary)
+      Log::Warn << "Loading '" << filename << "' as " 
+        << opts.FileTypeToString() << "; " 
+        << "but this may not be the actual filetype!" << std::endl;
 
-      return false;
-    }
+    success = matrix.load(stream, ToArmaFileType(opts.Format()));
+    if (!opts.NoTranspose())
+      inplace_trans(matrix);
+  }
+  return success;
+}
 
-    // There is still a small amount of differentiation that needs to be done:
-    // if we got a text type, it could be a coordinate list.  We will make an
-    // educated guess based on the shape of the input.
-    if (loadType == FileType::RawASCII)
+template <typename eT>
+bool LoadSparse(const std::string& filename,
+                arma::SpMat<eT>& matrix,
+                TextOptions& opts,
+                std::fstream& stream)
```

Comment:
This would be a good candidate for moving to a different file (same with `LoadDense`), just for the sake of organization, but, we don't have to do that now, up to you.

Reply:
You mentioned in chat that you'd do this in another PR; that sounds good to me so I'll resolve this conversation.